---
title: "Get more experience with R"
author: "Group XX"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: html_document
bibliography: COILbibliography.bib
---

# Introduction

In this R-challenge we would first like to reproduce a volcano plot from an article using the provided supplementary data and then apply some of the skills we have learned in the previous weeks. Basically, in the same article, the authors also provide a matrix with the expression of the proteins in the surfaceome of multiple myeloma cells. We will try to visualize this matrix as a heatmap, a correlation heatmap or a PCA/MDS plot.

__Article: The surfaceome of multiple myeloma cells suggests potential immunotherapeutic strategies and protein markers of drug resistance__
DOI: 10.1038/s41467-022-31810-6

The article "The surfaceome of multiple myeloma cells suggests potential immunotherapeutic strategies and protein markers of drug resistance" by Ferguson et al. presents a comprehensive investigation of the surface proteome, or surfaceome, of multiple myeloma cells. This study is pivotal as it explores how the surfaceome influences the interaction between tumor cells and their microenvironment, making it a critical area for potential therapeutic advancements and understanding drug resistance mechanisms [@Ferguson2022-sc].

The authors employed glycoprotein capture proteomics to systematically characterize the myeloma surfaceome under various conditions: at baseline, when facing drug resistance, and in response to acute drug treatments. This methodology allowed for a detailed examination of surface antigens that could be targeted for immunotherapy. The findings from this study include the establishment of a scoring system to evaluate and prioritize these surface antigens based on their potential therapeutic value [@Ferguson2022-sc].

One of the most significant outcomes of the study is the identification of CCR10 as a promising target for immunotherapeutic intervention. The expression of CCR10 was noted to be prevalent among malignant plasma cells, suggesting that therapies targeting this marker could enhance treatment efficacy. This provides a strategic direction for developing new immunotherapies aimed specifically at overcoming the challenges associated with drug resistance in multiple myeloma [@Ferguson2022-sc].

Overall, the article provides vital insights into the complexity of the myeloma surfaceome and reinforces the need for continued exploration into targeted therapies that can improve clinical outcomes in patients suffering from multiple myeloma.
*the article (pdf) is available in the folder of this tutorial*


## Libraries used for this tutorial:
You may not need all of these listed here.

- tidyr, dplyr: data handling from the tidyverse
- magrittr: together with tidyverse it allows the use of %&lt;&gt;%
- readxl and openxlsx: libraries to open and write microsoft excel files
- ggplot2: produce nice plots
- pheatmap: to great nice (pretty) heatmaps from matrices
- ggtree: to plot phylogenetic trees
- ape: to read and manipulate phylogenetic trees
- viridis: to get nice color scales
- DESeq2, plotly, limma, SummarizedExperiment: to generate similar QC plots as in week 2
- ggrepel: to avoid overplotting of labels in ggplot2

## Idea in this R-challenge:
* Find the excel sheet, look at it, see what is relevant and then try to do the coding
* One sheet in this excel book is relevant but only part of it, try to restructure it R the way you need it
* Try to use AI in this challenge at multiple steps (restructure the data, plotting w/ ggplot) if it is not immediately obvious to you how to do it.


```{r setup, include=FALSE}

if (! "BiocManager" %in% installed.packages()[,1]) {
  install.packages("BiocManager")
}

library(BiocManager)

# you may eventually need one of these packages.. therefore we directly install all of them
needed_packages <- c("tidyverse", "tidyr", "dplyr", "readxl", "ggplot2","openxlsx", 
                     "ggtree", "viridis", "ggrepel", "pheatmap", "SummarizedExperiment", 
                     "RColorBrewer", "magrittr", "DESeq2", "ape", "tibble", "plotly", "limma")

for (package in needed_packages) {
  if(! package %in% installed.packages()[,1]) {
    install(package, update = FALSE, ask=FALSE)
  }
  library(package, character.only = TRUE)
}
library(limma)
library(SummarizedExperiment)
library(plotly)
library(tibble)
library(magrittr)
library(DESeq2)
library(dplyr)
library(ggplot2)
library(tidyverse)
library(openxlsx)
library(ggtree)
library(ape)
library(viridis)
library(pheatmap)
library(ggrepel)
library(RColorBrewer)
```


# Goal 1: Redo Figure 1E from the article

Below here we show the Figure 1E from the article. We will try to reproduce this figure with the data provided in as supplement in an excel file.
_Try first to understand what this plot is showing. And explain it here, also indicate what columns are potentially relevant in this sheet (not all of them you need to use). Something like: on this volcano we see .. _ 
 
![NatCom: Ferguson et. al, Figure 1E](img/Fig1E.png)
```{r readInData, echo=FALSE, include=TRUE}
# read in the data and put it into an object (e.g. dat) have a look at the head of the data

dat <- read.xlsx("41467_2022_31810_MOESM3_ESM.xlsx", sheet = "Supplementary Dataset 1")
head(dat)

print("here we look into head of the data")

```


```{r reshapeData, echo=FALSE, include=TRUE}
# ideally you reshape the data into another object by selecting only necessairy columns
# maybe also rename the columns to make it easier to work with
vdat <- dat |> select(Protein, MM_vs_Leukemia_difference, MM_vs_Leukemia_pvalue)

# rename columns
colnames(vdat) <- c("Gene", "log2Ratio", "Pvalue")
head(vdat)
print("look into reshaped data") # these statements can also be commented afterwards (put a hash-tag infront of print)

```


```{r someFirstVolcano, echo=FALSE, include=TRUE}
# do a minimal volcano plot just to see if it works

# use ggplot to create volcano plot
ggplot(vdat, aes(x = log2Ratio, y = -log10(Pvalue))) +
  geom_point() +
  theme_minimal() +
  labs(x = "log2(MEAN.RATIO)", y = "-log10(Pvalue)")

print("Here our first minimal volcano plot")

```



```{r volcano With Significance colored, echo=FALSE, include=TRUE}
# usually proteins with a p-value < 0.05 and a log2FC > 1 (or -1) are considered significant

# define threshold for significance and use different color
abs_log2_threshold <- 1
pvalue_threshold <- 0.05
vdat <- vdat |> mutate(Significant = ifelse(abs(log2Ratio) > abs_log2_threshold & Pvalue < pvalue_threshold, "lightblue", "grey"))

# use the color column to color the points
ggplot(vdat, aes(x = log2Ratio, y = -log10(Pvalue), color = Significant)) +
  geom_point() +
  theme_minimal() +
  labs(x = "log2(MEAN.RATIO)", y = "-log10(Pvalue)")


```



```{r add some lables found in a file, echo=FALSE, include=TRUE}
# Now in the plot, some labels are added to some interesting proteins (POIs)

# read in another txt with the proteins of interest
membrane_proteins <- read.table("MarkerProteinsPOIs.txt", header = TRUE, sep = "\t")

vdat2 <- dat |> select(Uniprot_ID, Protein, MM_vs_Leukemia_difference, MM_vs_Leukemia_pvalue)

# rename columns
colnames(vdat2) <- c("UniprotIDs","Gene", "log2Ratio", "Pvalue")

# add vector for significance calls
vdat2 <- vdat2 |> mutate(Significant = ifelse(abs(log2Ratio) > abs_log2_threshold & Pvalue < pvalue_threshold, "lightblue", "grey"))

# add column w TRUE/FALSE if found in membrane_proteins$Silico.Surfaceome
vdat2$MP_bool <- vdat2$Gene %in%  membrane_proteins$MarkerProteins
print("How many of the list are matched in our volcano?: ")
sum(vdat2$MP_bool)

# do volcano plot and add labels if they are membrand proteins
ggplot(vdat2, aes(x = log2Ratio, y = -log10(Pvalue), color = Significant)) +
  geom_point() +
  geom_text(data = vdat2 |> filter(MP_bool), aes(label = Gene), nudge_y = 0.5) +
  theme_minimal() +
  labs(x = "log2(MEAN.RATIO)", y = "-log10(Pvalue)")



```


```{r some labels are overplotted can you do it better with ai, echo=FALSE, include=TRUE}

# Create the volcano plot with lines connecting points to labels
ggplot(vdat2, aes(x = log2Ratio, y = -log10(Pvalue), color = Significant)) +
    geom_point(size = 2) +  # Plot points
    # Add lines from points to labels
    geom_segment(data = vdat2 |> filter(MP_bool),
                 aes(x = log2Ratio, y = -log10(Pvalue),
                     xend = log2Ratio, yend = -log10(Pvalue) + 0.5),
                 color = "black", size = 0.5, alpha = 0.5) +  # Line to label
    # Add labels for membrane proteins (MP_bool should be a logical column)
    geom_text(data = vdat2 |> filter(MP_bool),
              aes(label = Gene),
              size = 3,
              color = "black",
              nudge_y = 0.5) +  # Adjust label position
    theme_minimal() +
    labs(x = "log2(MEAN.RATIO)", y = "-log10(Pvalue)") +
    scale_color_manual(values = c("grey", "blue")) +
    theme(
        legend.position = "top",
        axis.text = element_text(size = 12),
        axis.title = element_text(size = 14)
    )

```


```{r next version some labels are overplotted can you do it better with ai, echo=FALSE, include=TRUE}
# a nicer version of it with ggrepel
ggplot(vdat2, aes(x = log2Ratio, y = -log10(Pvalue), color = Significant)) +
    geom_point(size = 2) +  # Points
    geom_label_repel(data = vdat2 |> filter(MP_bool),
                     aes(label = Gene),
                     size = 3,
                     box.padding = 0.35,  # Padding around labels
                     point.padding = 0.5,  # Padding around the point and label
                     max.overlaps = 10,  # Limit how many labels can overlap
                     color = "black") +  # Label color
    theme_minimal() +
    labs(x = "log2(MEAN.RATIO)", y = "-log10(Pvalue)") +
    scale_color_manual(values = c("grey", "blue")) +
    theme(
        legend.position = "top",
        axis.text = element_text(size = 12),
        axis.title = element_text(size = 14)
    )

```

# Goal 2: Try to visualize the expression matrix as a heatmap and make some QC plots

The authors of the article also provide a matrix with the expression of the proteins in the surfaceome of multiple myeloma cells. 
Please also remember the QC-plots from week 3. We will try to visualize this matrix as a heatmap and do some more QC plots such as correlation heatmap or a PCA/MDS plot.


```{r look at excel sheet again and get matrix out, echo=TRUE, include=TRUE}
# Do you find a pattern that matches all the columns that you want?
dat_wide <- dat |> select(Protein, matches("\\.\\d+$"))
```
First some reshaping is necessairy again.

### A heatmap with clustering 

```{r pheatmap, echo=FALSE, include=TRUE}
# use the pheatmap package. It works on wide format
# Generate the heatmap with clustering (using pheatmap)
pheatmap(
    dat_wide %>% select(-Protein),  # Remove the 'Protein' column, which is now rownames
#    scale = "row",  # Optionally, scale rows (samples) if needed
    scale = "column",  # Optionally, scale rows (samples) if needed
    clustering_distance_rows = "euclidean",  # Euclidean distance for rows (samples)
    clustering_distance_cols = "euclidean",  # Euclidean distance for columns (proteins)
    clustering_method = "complete",  # Complete linkage clustering
    show_rownames = FALSE,  # Display row names (samples)
    show_colnames = TRUE,  # Display column names (proteins)
    color = viridis::viridis(100)  # Use a viridis color palette
)


```

### A Correlation-Heatmap

```{r correlation plot or pca, echo=FALSE, include=TRUE}
# corr
cor_mat <- cor(dat_wide %>% select(-Protein))
pheatmap(cor_mat, color = viridis::viridis(100))
```

## Some QC plots like in week 3

### First again some reshaping
Here we take the recipe from week 3 and apply it to the data we have here.
Important here is that the __protein names__ are redundant that causes problems. 
Therefore we take the Uniprot_ID instead of the protein names to generate the SummarizedExperiment object.

```{r qc plots ala week3 prepare it, echo=TRUE, include=TRUE, warning=FALSE}
# extract again relevant columns but here UniProt Ids (since unique) for all rows
dat_wide2 <- dat |> select(Uniprot_ID, matches("\\.\\d+$"))
# only quantitative matrix in a data frame and uniprot Ids as rownames
qMat <- dat_wide2 %>% select(-Uniprot_ID)
rownames(qMat) <- dat_wide$UniProt_ID
# qMat should have integer values to minimize rounding effect errors we multiply it with 100
qMat100 <- round(qMat*100, 0)


# extract conditions (cell types) from column names with regular expression
cellTypes <- gsub("\\.\\d+$", "", colnames(qMat100))

# Define meta dataframe for later use
meta <- data.frame(
  Condition=as.factor(cellTypes),
  row.names=colnames(qMat100)
)

conditionColours <- scales::hue_pal()(length(unique(meta$Condition)))
names(conditionColours) <- unique(meta$Condition)
sampleColours <- conditionColours[meta$Condition]


dds <- DESeq2::DESeqDataSetFromMatrix(countData=qMat100,
                                      colData=meta,
                                      design=~Condition)
vsd <- DESeq2::vst(dds)

# Extract normalized counts
vsdSE <- SummarizedExperiment::assay(vsd)
```
### PCA and MDS plots like in Week 3

```{r PCA plot, echo=TRUE, include=TRUE, warning=FALSE}
# Run PCA
pcDat  <- prcomp(t(vsdSE), scale. = FALSE)

# Calculate explained variance
varExp <- (100*pcDat$sdev^2)/sum(pcDat$sdev^2)

# Store the explained variance of top 8 PCs
varExp_df <- data.frame(PC= paste0("PC",1:8),
                          varExp=varExp[1:8])

# Scree plot
# varExp_df %>%
#   ggplot(aes(x=PC,y=varExp, group=1)) +
#   geom_point(colour="steelblue", size=4) +
#   geom_col(fill="steelblue") +
#   geom_line() + 
#   theme_bw() + ylim(c(0,100))

#PCA plot
plot_ly(as.data.frame(pcDat$x), x=~PC1, y=~PC2, color=meta$Condition, colors="Set1",
        type="scatter", mode="markers") %>%
  layout(title="PCA Plot")
```

```{r mds plot, echo=TRUE, include=TRUE, warning=FALSE}
mds <- limma::plotMDS(vsdSE, plot=FALSE)
mdsOut <- mds$eigen.vectors[,1:3]
colnames(mdsOut) <- c("Leading logFC dim1", "Leading logFC dim2", 
                      "Leading logFC dim3")
toPlot <- cbind(meta %>% rownames_to_column("Sample"), mdsOut)
plot_ly(toPlot, x=~`Leading logFC dim1`, y=~`Leading logFC dim2`, z=~`Leading logFC dim3`, color=~Condition, colors="Set1", type='scatter3d', mode='markers+text', text=~Sample, textposition = "top right") %>%
  plotly::layout(title="Classical MDS", scene=list(xaxis=list(title = 'Leading logFC dim1'), yaxis = list(title = 'Leading logFC dim2'), zaxis = list(title = 'Leading logFC dim3')))
```

### Correlation plot like in week 3

```{r corrPlot like in week 3, echo=TRUE, include=TRUE, warning=FALSE}
# Pearson correlation plot 
pheatmap(
  mat               = cor(vsdSE, use="complete.obs"),
  treeheight_row    = 100,
  treeheight_col    = 100,
  cutree_rows       = 2, 
  cutree_cols       = 2,
  silent            = F,
  annotation_col    = meta,
  annotation_colors = list(Condition = conditionColours),
  color             = brewer.pal(n = 9, name = "Blues"),
  fontsize_row      = 4, 
  fontsize_col      = 4,
  display_numbers   = TRUE,
  fontsize_number   = 4)
```

# Session Information

Here we show the session information of the packages used in this tutorial as well as the references can be found below.

```{r sessionInfo, echo=FALSE, include=TRUE}
pander::pander(sessionInfo())
```


# References




